src/dotnet/GeneratorAgent/src/Azure.Tools.GeneratorAgent/GithubFilesRetrieveService.cs

```diff
@@ -0,0 +1,152 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Azure.Tools.GeneratorAgent.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Azure.Tools.GeneratorAgent
+{
+    /// <summary>
+    /// TypeSpec file service for GitHub repository access using GitHub API for fast file retrieval.
+    /// </summary>
+    internal class GitHubFilesRetrieveService : IDisposable
+    {
+        private readonly ILogger<GitHubFilesRetrieveService> Logger;
+        private readonly AppSettings AppSettings;
+        private readonly HttpClient HttpClient;
+        private readonly string CommitId;
+        private readonly string TypespecSpecDir;
+
+        // Static JSON options for better performance (reused across calls)
+        private static readonly JsonSerializerOptions JsonOptions = new()
+        {
+            PropertyNameCaseInsensitive = false,
+            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+        };
+
+        public GitHubFilesRetrieveService(
+            AppSettings appSettings,
+            ILogger<GitHubFilesRetrieveService> logger,
+            ValidationContext validationContext)
+        {
+            ArgumentNullException.ThrowIfNull(appSettings);
+            ArgumentNullException.ThrowIfNull(logger);
+            ArgumentNullException.ThrowIfNull(validationContext);
+
+            AppSettings = appSettings;
+            Logger = logger;
+
+            CommitId = validationContext.ValidatedCommitId;
+            TypespecSpecDir = validationContext.ValidatedTypeSpecDir;
+
+            // Initialize HttpClient with User-Agent header and timeout for GitHub API
+            HttpClient = new HttpClient
+            {
+                Timeout = TimeSpan.FromMinutes(2) // Reasonable timeout for file downloads
+            };
+            HttpClient.DefaultRequestHeaders.Add("User-Agent", "AzureSDK-TypeSpecGenerator/1.0");
```

Comment:
Per Jesse's comment about not making a new HttpClient each time, you could make this static shared like in https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/Pipeline/HttpClientPipelineTransport.cs#L19